src/monodroid/jni/monodroid-glue.cc

```diff
@@ -2551,6 +2557,29 @@ MonodroidRuntime::get_java_class_name_for_TypeManager (jclass klass)
 	return ret;
 }
 
+void
+MonodroidRuntime::monodroid_javasystem_loadLibrary (const char *libname) noexcept
+{
+	JNIEnv *env = osBridge.ensure_jnienv ();
+	jmethodID loadLibraryID = env->GetStaticMethodID (mono_android_Runtime, "loadLibrary", "(Ljava/lang/String;)Z");
+	if (loadLibraryID == nullptr) {
+		log_fatal (LOG_DEFAULT, "Failed to look up the mono.android.Runtime.loadLibrary Java method");
+		Helpers::abort_application ();
+	}
+
+	jstring libname_java = env->NewStringUTF (libname);
+	jboolean success = env->CallStaticBooleanMethod (mono_android_Runtime, loadLibraryID, libname_java);
+	env->DeleteLocalRef (libname_java);
+
+	if (success != 0) {
+		// Loaded
+		return;
+	}
+
+	log_warn (LOG_DEFAULT, "System.loadLibrary failed for '%s', attempting to load the library directly", libname);
+	monodroidRuntime.monodroid_dlopen (libname, MONO_DL_LOCAL, nullptr, nullptr);
```

Comment:
Can this fail?  (Presumably it can/should, e.g. when someone does `JavaSystem.LoadLibrary("does-not-exist")`.

If it does fail, what should happen?

It feels like `JavaSystem.LoadLibrary()` should re-raise the *original* Java-side exception if this call fails, and to do *that* we'd have to rework things, e.g. have `Runtime.loadLibrary()` return a `java.lang.Throwable` instance and use `JNIEnv::CallStaticObjectMethod()` to invoke it.

What should happen *on success*?  The original report was that `JNI_OnLoad()` wasn't invoked.  Will `monodroid_dlopen()` invoke `JNI_OnLoad()`?  If not, should *we* look for and invoke it?

Reply:
If that load fails, then the attempt to use the library's interfaces will cause an error later on in the application. I thought that trying to throw a **Java** exception from here and in this situation would be quite complicated and, frankly, not worth the effort.  Also aborting the application for failure to find the library would be, I think, not desired here.  Letting the application fail because of the missing library, I think, is a fair compromise, but of course not ideal.